src/modules/project-config.ts

```diff
@@ -84,35 +96,124 @@ export class ProjectConfig implements ProjectConfigOptions {
             packageConfig.version = latestVersion;
-            projectConfig.packages.push(packageConfig);
+            projectConfig.getPackages().push(packageConfig);
         }
-        projectConfig.variables.set('language', language);
-        projectConfig.variables.set('repoType', 'app');
-        projectConfig.variables.set('appManifestPath', DEFAULT_APP_MANIFEST_PATH);
-        projectConfig.variables.set('githubRepoId', '<myrepo>');
+        projectConfig.getVariableMappings().set('language', language);
+        projectConfig.getVariableMappings().set('repoType', 'app');
+        projectConfig.getVariableMappings().set('appManifestPath', DEFAULT_APP_MANIFEST_PATH);
+        projectConfig.getVariableMappings().set('githubRepoId', '<myrepo>');
         projectConfig.write();
     }
 
+    /**
+     * Write the project configuration to file.
+     *
+     * @param path Path of the file to write the configuration to.
+     */
     write(path: PathLike = DEFAULT_CONFIG_FILE_PATH): void {
-        // replaceAll because of having "backward compatibility" before deprecate "name" completely
-        const configString = `${JSON.stringify(this, mapReplacer, 4).replaceAll('"repo"', '"name"')}\n`;
+        // if we find an "old" project configuration with no components explicitly mentioned
+        // we persist all components we can find.
+        let componentsToSerialize: ComponentConfig[] = this._components;
+
+        if (!componentsToSerialize || componentsToSerialize.length === 0) {
+            componentsToSerialize = this.getComponents().map((cc) => cc.config);
+        }
+
+        const projectConfigOptions: ProjectConfigOptions = {
+            packages: this._packages,
+            components: componentsToSerialize,
+            variables: this._variables,
+            cliVersion: this.cliVersion,
+        };
+        const configString = `${JSON.stringify(projectConfigOptions, mapReplacer, 4)}\n`;
         writeFileSync(path, configString, DEFAULT_BUFFER_ENCODING);
     }
+
+    /**
+     * Return the configuration of a component.
+     *
+     * @param projectConfig The project configuration.
+     * @param componentId   The ID of the component.
+     * @returns The configuration of the component.
+     */
+    getComponentConfig(componentId: string): ComponentConfig {
+        var maybeComponentConfig: ComponentConfig | undefined;
+        if (this._components) {
+            maybeComponentConfig = this._components.find((c) => c.id === componentId);
+        }
+        return maybeComponentConfig ? maybeComponentConfig : new ComponentConfig(componentId);
+    }
+
+    /**
+     * Return all components used by the project. If the project specifies no components explicitly,
+     * all components are used by default.
+     *
+     * @returns A list of all components used by the project.
+     */
+    getComponents(): Array<ComponentContext> {
+        var componentContextes = new Array<ComponentContext>();
```

Comment:
let over var and better to type "let componentContextes: ComponentContext[] = []"